tests/Aql/AsyncExecutorTest.cpp

```diff
@@ -0,0 +1,198 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2023-2023 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Tobias Gödderz
+////////////////////////////////////////////////////////////////////////////////
+
+#include <gtest/gtest.h>
+
+#include "Aql/AqlExecutorTestCase.h"
+#include "Mocks/FakeScheduler.h"
+
+#include "Aql/AsyncExecutor.h"
+
+#include <random>
+
+using namespace arangodb;
+using namespace arangodb::aql;
+using namespace arangodb::tests;
+
+struct AsyncExecutorTest : AqlExecutorTestCase<false> {
+  AsyncExecutorTest()
+      : AqlExecutorTestCase(&scheduler), scheduler(_server->server()) {}
+
+  FakeScheduler scheduler;
+};
+
+// Regression test for https://arangodb.atlassian.net/browse/BTS-1325.
+// See https://github.com/arangodb/arangodb/pull/18729 for details.
+TEST_F(AsyncExecutorTest, sleepingBeauty) {
+  auto registerInfos = RegisterInfos(RegIdSet{}, RegIdSet{}, 1, 1,
+                                     RegIdFlatSet{}, RegIdFlatSetStack{{0}});
+
+  auto testHelper = makeExecutorTestHelper();
+  testHelper
+      .addDependency<AsyncExecutor>(registerInfos, {}, ExecutionNode::ASYNC)
+      .addDependency<AsyncExecutor>(registerInfos, {}, ExecutionNode::ASYNC)
+      .setInputFromRowNum(1)
+      .setWaitingBehaviour(WaitingExecutionBlockMock::WaitingBehaviour::ALWAYS)
+      .setCall(AqlCall{0u, AqlCall::Infinity{}, AqlCall::Infinity{}, false});
+
+  auto* asyncBlock0 = dynamic_cast<ExecutionBlockImpl<AsyncExecutor>*>(
+      testHelper.pipeline().get().at(0).get());
+  // Having the nodes in a certain order (i.e. pipeline[0].id() == 0, and
+  // pipeline[1].id() == 1), makes reading profiles less confusing.
+  ASSERT_EQ(asyncBlock0->getPlanNode()->id().id(), 0);
+
+  // one initial "wakeup" to start execution
+  auto wakeupsQueued = 1;
+  auto wakeupHandler = [&wakeupsQueued]() noexcept {
+    ++wakeupsQueued;
+    return true;
+  };
+  testHelper.setWakeupHandler(wakeupHandler);
+  testHelper.setWakeupCallback(wakeupHandler);
+  testHelper.prepareInput();
+
+  asyncBlock0->setPostAsyncExecuteCallback([&] {
+    while (!scheduler.queueEmpty()) {
+      scheduler.runOnce();
+    }
+    while (wakeupsQueued > 0) {
+      --wakeupsQueued;
+      testHelper.executeOnce();
+    }
+  });
```

Comment:
I think this deserves a comment what we are doing here and why. Something like "the postAsyncExecuteCallback is called after the AsyncNode has called execute on its dependency, but before the results of that call is stored, so the node is still "inProgress". By making sure all all tasks an wakeups are processed before leaving the callback we simulate the thread being slow."